### PR TITLE
Lots of wrench fixes and features

### PR DESCRIPTION
**Pull request by vvuk**
Support for many new yaml display items, clips, and other bits.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/660)

### CODE REVIEW

File: webrender_traits/src/display_item.rs [hunk 1]
```diff
@@ -48,6 +48,27 @@ impl BorderRadius {
...
         }
     }
+
+    pub fn is_uniform(&self) -> Option<f32> {
+        let uniform_radius = LayoutSize::new(self.top_left.width, self.top_left.width);
+        if self.top_left == uniform_radius &&
+           self.top_right == uniform_radius &&
+           self.bottom_left == uniform_radius &&
+           self.bottom_right == uniform_radius
+        {
```
Comment (glennw) [inline webrender_traits/src/display_item.rs]: nit: brace placement

File: webrender_traits/src/display_item.rs [hunk 2]
```diff
@@ -48,6 +48,27 @@ impl BorderRadius {
             bottom_right: LayoutSize::new(radius, radius),
         }
     }
+
+    pub fn is_uniform(&self) -> Option<f32> {
+        let uniform_radius = LayoutSize::new(self.top_left.width, self.top_left.width);
+        if self.top_left == uniform_radius &&
```
Comment (glennw) [inline webrender_traits/src/display_item.rs]: nit: don't need to check top_left here.

**Comment by vvuk** [comment]
Fixed.

File: wrench/src/main.rs [hunk 3]
```diff
@@ -161,28 +164,48 @@ fn main() {
...
     for event in window.wait_events() {
+        if let Some(window_title) = wrench.take_title() {
+            window.set_title(&window_title);
+        }
+
         if let Some(limit) = limit_seconds {
             if (time::SteadyTime::now() - time_start) >= limit {
+                let mut block_avg_ms = block_avg_ms.iter().map(|v| as_ms(*v)).collect::<Vec<f64>>();
+                block_avg_ms.sort_by(|a, b| a.partial_cmp(b).unwrap());
+                let avg_ms = block_avg_ms.iter().fold(0., |sum, v| sum + v) / block_avg_ms.len() as f64;
```
Comment (kvark) [inline wrench/src/main.rs]: nit: could use [sum](https://doc.rust-lang.org/std/iter/trait.Sum.html#tymethod.sum)

File: wrench/src/wrench.rs [hunk 4]
```diff
@@ -174,39 +179,50 @@ impl Wrench {
...
-
+            gl_renderer: gl_renderer,
+            gl_version: gl_version,
             frame_start_sender: timing_sender,
         };
 
         wrench.set_title("start");
+        // there's a "frame 0" that webrender itself renders; push this to
+        // not confuse our notifier
+        wrench.frame_start_sender.push(time::SteadyTime::now());
```
Comment (kvark) [inline wrench/src/wrench.rs]: nit: use `wrench.begin_frame()` instead?
perhaps, `frame_start_sender` doesn't even need to be public

File: wrench/src/yaml_frame_reader.rs [hunk 5]
```diff
@@ -66,23 +94,176 @@ impl YamlFrameReader {
         self.add_stacking_context_from_yaml(wrench, &yaml["root"]);
     }
 
+    fn to_clip_region(&mut self, item: &Yaml, item_bounds: &LayoutRect, wrench: &mut Wrench) -> Option<ClipRegion> {
```
Comment (kvark) [inline wrench/src/yaml_frame_reader.rs]: sweeet!

File: wrench/src/yaml_frame_reader.rs [hunk 6]
```diff
@@ -66,23 +94,176 @@ impl YamlFrameReader {
...
 
-    fn handle_image(&mut self, wrench: &mut Wrench, clip_region: &ClipRegion, item: &Yaml)
+    fn handle_gradient(&mut self, wrench: &mut Wrench, clip_region: &ClipRegion, item: &Yaml)
+    {
+        let bounds_key = if item["type"].is_badvalue() { "gradient" } else { "bounds" };
+        let bounds = item[bounds_key].as_rect().expect("gradient must have bounds");
+        let start = item["start"].as_point().expect("gradient must have start");
+        let end = item["end"].as_point().expect("gradient must have end");
+        let stops = if let Some(stops) = item["stops"].as_vec() {
+            // FIXME(vlad): I can't find the right way to do this with iterators;
```
Comment (kvark) [inline wrench/src/yaml_frame_reader.rs]: Chunks come to rescue:
```rust
let g_stops = stops.chunks(2).map(|chunk| GradientStop {
  offset: chunk[0].as_force_f32().expect("gradient stop offset is not f32"),
  color: chunk[1].as_colorf().expect("gradient stop color is not color"),
}).collect::<Vec<_>>();
```

File: wrench/src/yaml_frame_reader.rs [hunk 7]
```diff
@@ -66,23 +94,176 @@ impl YamlFrameReader {
...
+                Some(self.builder().new_clip_region(&bounds, complex, image_mask))
+            }
+            _ => {
+                None
+            }
+        }
+    }
+
     fn handle_rect(&mut self, wrench: &mut Wrench, clip_region: &ClipRegion, item: &Yaml)
     {
```
Comment (kvark) [inline wrench/src/yaml_frame_reader.rs]: opening brackets on the new line is against the rust style guidelines

File: wrench/src/yaml_frame_reader.rs [hunk 8]
```diff
@@ -66,23 +94,176 @@ impl YamlFrameReader {
...
+
+    fn handle_border(&mut self, wrench: &mut Wrench, clip_region: &ClipRegion, item: &Yaml)
+    {
+        let bounds_key = if item["type"].is_badvalue() { "border" } else { "bounds" };
+        let bounds = item[bounds_key].as_rect().expect("borders must have bounds");
+        let widths = item["width"].as_vec_f32().expect("borders must have width(s)");
+        let colors = item["color"].as_vec_colorf().expect("borders must have color(s)");
+        let styles = item["style"].as_vec_string().expect("borders must have style(s)");
+        let styles = styles.iter().map(|s| match s {
+            s if s == "none" => BorderStyle::None,
```
Comment (kvark) [inline wrench/src/yaml_frame_reader.rs]: You should be able to just do `"something" => BorderStyle::Something` without all the `if` in the matching

File: wrench/src/yaml_frame_reader.rs [hunk 9]
```diff
@@ -66,23 +94,176 @@ impl YamlFrameReader {
...
+        let bounds = item[bounds_key].as_rect().expect("box shadow must have bounds");
+        let box_bounds = item["box_bounds"].as_rect().unwrap_or(bounds);
+        let offset = item["offset"].as_point().unwrap_or(TypedPoint2D::zero());
+        let color = item["color"].as_colorf().unwrap_or(ColorF::new(0.0, 0.0, 0.0, 1.0));
+        let blur_radius = item["blur_radius"].as_force_f32().unwrap_or(0.0);
+        let spread_radius = item["spread_radius"].as_force_f32().unwrap_or(0.0);
+        let border_radius = item["border_radius"].as_force_f32().unwrap_or(0.0);
+        let clip_mode = if let Some(mode) = item.as_str() {
+            match mode {
+                s if s == "none" => BoxShadowClipMode::None,
```
Comment (kvark) [inline wrench/src/yaml_frame_reader.rs]: same here. Just make sure you are matching `&str` and not something else

File: wrench/src/yaml_frame_writer.rs [hunk 10]
```diff
@@ -14,6 +13,7 @@ use std::path::{Path, PathBuf};
 use webrender;
 use webrender_traits::*;
 use yaml_rust::{Yaml, YamlEmitter};
+use time;
 
 use super::CURRENT_FRAME_NUMBER;
```
Comment (kvark) [inline wrench/src/yaml_frame_writer.rs]: any reason we can't just store the frame number locally and pass it around?

**Comment by kvark** [comment]
Thanks @vvuk ! I think it's in good shape now. There are still minor issues to be addressed, but we can do it as a follow up, especially considering that some one is blocked by this PR (see #669).
@glennw - if you happy and you know it... please merge

**Pull request merged**